### PR TITLE
Feature/device-annotation

### PR DESCRIPTION
**Pull request by WingsOfPanda**
The purpose of this pr is to replicate `colocate_gradients_with_ops` option to `tf.gradients` in 1.x to 2.x.

### CODE REVIEW

**Comment by WingsOfPanda** [comment]
please refer to this [https://github.com/tensorflow/tensorflow/issues/54460](url) feature request post for more details

File: tensorflow/core/public/version.h [hunk 1]
```diff
@@ -26,7 +26,7 @@ limitations under the License.
 
 // TF_VERSION_SUFFIX is non-empty for pre-releases (e.g. "-alpha", "-alpha.1",
 // "-beta", "-rc", "-rc.1")
-#define TF_VERSION_SUFFIX ""
+#define TF_VERSION_SUFFIX "-device-annotation-mkl-support"
```
Comment (cantonios) [inline tensorflow/core/public/version.h:29]: This definitely shouldn't be changed.  If anything, we could potentially check for a previous definition

e.g.
```
#ifndef TF_VERSION_SUFFIX
#define TF_VERSION_SUFFIX ""
#endif
```

Comment (WingsOfPanda) [reply tensorflow/core/public/version.h]: sorry my bad. just need to suffix sth so I know what features that I added into my personal build. not intended to merge into the upstream... I already removed this

Comment (cantonios) [reply tensorflow/core/public/version.h:29]: Looks like this is still here, needs to be removed.

File: tensorflow/python/eager/pywrap_tfe_src.cc [hunk 2]
```diff
@@ -3357,6 +3380,7 @@ PyObject* RecordGradient(PyObject* op_name, PyObject* inputs, PyObject* attrs,
       forward_function);
 
   Py_DECREF(num_inputs);
+//  Py_DECREF(device);
```
Comment (cantonios) [inline tensorflow/python/eager/pywrap_tfe_src.cc:3383]: Why is this here?

**Review by cantonios — CHANGES_REQUESTED** [review]

**Review by sagunb — COMMENTED** [review]
Thanks for the PR! Overall looks good! Please address the changes requested by cantonios@

**Comment by gbaned** [comment]
Hi @WingsOfPanda Any update on this PR? Please. Thank you!

**Comment by mihaimaruseac** [comment]
Closing as it stalled

**Comment by WingsOfPanda** [comment]
@mihaimaruseac hi sir. could you re-open this pr? sorry I was 'lost' for awhile. Now I'm ready to fix the problem mentioned above!

**Comment by gbaned** [comment]
Hi @WingsOfPanda Any update on this PR? Please. Thank you!
